packages/react-devtools-extensions/src/generateHookMap.js

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import traverse, {Node, NodePath} from '@babel/traverse';
+import {string} from 'yargs';
+import {getHookNamesFromAST} from './astUtils';
+import {encode, decode} from 'sourcemap-codec';
+
+export type HookMap = {|
+  names: $ReadOnlyArray<string>,
+  mappings: $ReadOnlyArray<
+    $ReadOnlyArray<[number, number, number, number, number]>,
+  >,
+|};
+
+export type EncodedHookMap = {|
+  names: $ReadOnlyArray<string>,
+  mappings: string,
+|};
+
+export function generateHookMap(sourceAST: Node): HookMap {
+  const hooksFromAST = getHookNamesFromAST(sourceAST);
+
+  const names = [];
+  const mappings = [];
+  hooksFromAST.forEach((hookNodes, name) => {
+    names.push(name);
+    hookNodes.forEach(node => {
+      mappings.push([
+        [
+          node.loc.start.line,
+          node.loc.start.column,
+          node.loc.end.line,
+          node.loc.end.column,
+          names.length - 1,
+        ],
+      ]);
+    });
+  });
```

Comment:
For lookup to be a cheap binary search, mappings need to have a well-defined sort order (based on `node.loc.start` probably?) and they should not be overlapping/nested. Is this guaranteed by the traversal logic? Worth documenting in a comment.

Note that if we did need to support nesting (e.g. maybe a `useMemo` call that returns a component function?), we could do something that works more like Metro's function maps: keep a stack while generating the list and record a (single) position in the list whenever the top of the stack changes. The main difference is that for hook maps we'd need a sentinel "no hook" value, whereas function maps always have at least the `<global>` function.

Reply:
ah makes sense, thanks for the extra context here. 

---

> Is this guaranteed by the traversal logic? Worth documenting in a comment.

I believe so, but I will add some tests to verify as well.

--- 
> Note that if we did need to support nesting (e.g. maybe a useMemo call that returns a component function?)

I assume you mean something like this?

```
const comp = useMemo(() => () => {
  const [state, setState] = useState();
  return (...);
});
```

that's a good point, I hadn't actually considered this possibility, which is why I opted for the simpler mappings. I think it makes sense to support this (cc @bvaughn).

Reply:
Built-in hooks can't be nested. This is not supported by React.

Built-in hooks can be nested within custom hooks though, e.g.

```js
function useSomeThing() {
  const [foo, setFoo] = useState();

  // ...
}
```